XenModel/Actions/Updates/DownloadAndUpdateClientAction.cs

```diff
@@ -0,0 +1,292 @@
+
+                    if (updateDownloadState == DownloadState.Error)
+                    {
+                        needToRetry = true;
+
+                        // this many errors so far - including this one
+                        errorCount++;
+
+                        // logging only, it will retry again.
+                        log.ErrorFormat(
+                            "Error while downloading from '{0}'. Number of errors so far (including this): {1}. Trying maximum {2} times.",
+                            address, errorCount, MAX_NUMBER_OF_TRIES);
+
+                        if (updateDownloadError == null)
+                            log.Error("An unknown error occurred.");
+                        else
+                            log.Error(updateDownloadError);
+                    }
+                } while (errorCount < MAX_NUMBER_OF_TRIES && needToRetry);
+            }
+            finally
+            {
+                //deregister download events
+                client.DownloadProgressChanged -= client_DownloadProgressChanged;
+                client.DownloadFileCompleted -= client_DownloadFileCompleted;
+
+                NetworkChange.NetworkAvailabilityChanged -= NetworkAvailabilityChanged;
+
+                client.Dispose();
+            }
+
+            //if this is still the case after having retried MAX_NUMBER_OF_TRIES number of times.
+            if (updateDownloadState == DownloadState.Error)
+            {
+                log.ErrorFormat("Giving up - Maximum number of retries ({0}) has been reached.", MAX_NUMBER_OF_TRIES);
+
+                MarkCompleted(updateDownloadError ?? new Exception(Messages.ERROR_UNKNOWN));
+            }
+
+        }
+
+        private void NetworkAvailabilityChanged(object sender, NetworkAvailabilityEventArgs e)
+        {
+            if (!e.IsAvailable && client != null && updateDownloadState == DownloadState.InProgress)
+            {
+                updateDownloadError = new WebException(Messages.NETWORK_CONNECTIVITY_ERROR);
+                updateDownloadState = DownloadState.Error;
+                client.CancelAsync();
+            }
+        }
+
+        protected override void Run()
+        {
+            if (downloadUpdate)
+            {
+                log.InfoFormat("Downloading '{0}' installer (from '{1}') to '{2}'", updateName, address, outputPathAndFileName);
+                Description = string.Format(Messages.DOWNLOAD_CLIENT_INSTALLER_ACTION_DESCRIPTION, updateName);
+                LogDescriptionChanges = false;
+                DownloadFile();
+                LogDescriptionChanges = true;
+
+                if (IsCompleted || Cancelled)
+                    return;
+
+                if (Cancelling)
+                    throw new CancelledException();
+            }
+            
+            ValidateMsi();
+
+            // Install the msi            
+            try
+            {
+                if (File.Exists(outputPathAndFileName))

```

Comment:
Should we error out and/or log at this point if the `outputPathAndFileName` wasn't found?